### DISCUSSION TITLE
Breaking change of dicomdataset Comparing

### DISCUSSION THREAD

**Discussion by Tony-Jay** [fo-dicom/fo-dicom#1787] [General]
Hi,
I am unable to update FO-dicom   5.1.2. since a recent update which add a comparer for datasets

I assume it is from "Added IEquatable implementation and equality operators for DicomDataset class"

This seems to affect us where previously references were used in explicit dictionaries and implicit hashsets in LINQ queries.

I am assuming that when comparing is going on in these collections a huge object is created to keep the item used to compare - and the object may stay around for the life of the dataset ?

This explodes our memory use to 10s of GB for our process  (leading to out of memory exceptions ) compared to 100s of MB and it much slower  when it does not run out of memory.

I assume we can make our code compliant by using our own comparer, but care is need to spot where LINQ is creating hash sets.

I would have expected this to be listed under the breaking changes section (which it is not) and even a more major version increase due to impact.

Others may be affected in a similar way, but it could be more subtle for them.

Would a global option to enable the previous behaviour a good migration path ?

Thoughts ? 

Tony

**Comment by mrbean-bremen** [comment]
Cc @ViktoriaSalociova

**Comment by Tony-Jay** [comment]
@gofal  do you have an opinion on this as we currently are not updating to the latest version due to the number of breaking changes it has introduced (and the care needed now when using a link statement)

**Comment by gofal** [comment]
I try to be cautous not to intoduce breaking changes or risks without explicitly announcing it.
That adding an IEquatable might cause such an issue surprises me. 

I did some benchmarks: the calculation of the Hashcode should not be a problem, the code is super fast (so 10 to 15 Microseconds) and allocates no memory. 
But the comparision itself, when you code `dataset1 == dataset2` or `dataset1.Equals(dataset2)` then now this takes some time and memory, because the content of the tags are compared. I cannot see that the memory is persistet somewhere, but the memory is allocated and it stays until the garbage collector has time to cleanup.

**Comment by gofal** [comment]
DicomDataset inherits from IEquatable, this is something you cannot change with a blobal option. But the DicomDatasetComparer, that is used under the hood, could have some global options like "Only compare with ReferenceEquals" or "Compare all content".
In the PR #1668 there was posted, that it is up to following stories to allow more options for the DicomDatasetComparer.

Please open an issue where we can address this and add some global options, before we do the next release.

**Comment by gofal** [comment]
created issue #1807

**Reply by gofal** [reply]
@Tony-Jay  please review the benchmarks that I posted in the issue. Do you thing, you still need the option to turn the content-comparison off?

**Reply by Tony-Jay** [reply]
@gofal actually my request would be go back to the old behaviour and let's people enable non reference results and a new option

What we saw was  massively increased memory use (from 100'smb to 16GB when the paging file had issues)  when using dictionaries and LINQ, so may be I need to make a benchmark to show that ?
I think the problem with linq it the hashsets it makes under the hood, so it is not intuitive what problems you can get as collections grow
